### ISSUE TITLE
Cached catalogs + hiera-puppet-helper

### ISSUE THREAD

**Issue by barrkel-m20** [rodjek/rspec-puppet#215]
I noticed, after several hours of debugging, that catalogs appear to be cached in a class variable and end up being reused across multiple Rspec examples even though parameterization data (`let(:hiera_data) {}` via https://github.com/mthibaut/hiera-puppet-helper) has changed.

This line does the caching:

https://github.com/rodjek/rspec-puppet/blob/389f99ef666521fec1b4530fe69dc1ab84a060a8/lib/rspec-puppet/support.rb#L162

Why is this? It doesn't appear to be responsible for a major speedup, as far as I can tell, yet has major potential for test pollution.

The symptom is that later run examples get hiera data from earlier examples.

**Comment by shulima** [comment]
I was trying to figure out why enabling hiera lookup through shared context with metadata wasn't working as intended. Turned out that depending on test order, catalog cached from the hiera-enabled test was polluting the tests that were specifically supposed not to see hiera. Thank you for the extremely helpful workaround.

**Comment by jhoblitt** [comment]
Kudos to @ngiger !

**Comment by pall-valmundsson** [comment]
I came across this issue not because catalog reuse but because of memory consumption of the cache. 
The above modifications have performance issues when running many examples against the same catalog as the catalog has to be recompiled for each example (it's noticeable for large catalogs). My version tries to re-use the cache for serially run test cases but purges the cache as soon as new arguments are sent to build_catalog.

This is not heavily tested :)

```
module RSpec::Puppet
  module Support
    def build_catalog(*args)
      if @@cache.has_key?(args)
        @@cache[args]
      else
        @@cache = {}
        @@cache[args] ||= self.build_catalog_without_cache(*args)
      end
    end
  end
end
```

**Comment by domcleal** [comment]
@pall-valmundsson I've hopefully addressed the memory issue in #329 by limiting the number of cached catalogues, while still keeping multiple in memory.

The Hiera issue here's still valid I think, the only invalidation at the moment is based on :hiera_config, not :hiera_data.
